encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/ReflectiveObjectEncoderProvider.java

```diff
@@ -63,26 +68,48 @@ public void encode(@Nullable Object obj, @NonNull ObjectEncoderContext ctx) thro
   @NonNull
   @Override
   public <T> ObjectEncoder<T> get(@NonNull Class<T> type) {
-    Map<String, EncodingDescriptor> fields = new HashMap<>();
+    Map<FieldDescriptor, EncodingDescriptor> fields = new LinkedHashMap<>();
     for (Method method : type.getMethods()) {
       if (method.isAnnotationPresent(Encodable.Ignore.class)) {
         continue;
       }
 
       Encodable.Field fieldAnnotation = method.getAnnotation(Encodable.Field.class);
-      String getter = toGetterName(method, fieldAnnotation);
-      if (getter != null) {
-        method.setAccessible(true);
-        fields.put(
-            getter,
-            new EncodingDescriptor(fieldAnnotation != null && fieldAnnotation.inline(), method));
+      FieldDescriptor descriptor = toFieldDescriptor(method, fieldAnnotation);
+      if (descriptor == null) {
+        continue;
       }
+      method.setAccessible(true);
+
+      fields.put(
+          descriptor,
+          new EncodingDescriptor(fieldAnnotation != null && fieldAnnotation.inline(), method));
     }
     @SuppressWarnings("unchecked")
     ObjectEncoder<T> encoder = (ObjectEncoder<T>) new ReflectiveObjectEncoderImpl(fields);
     return encoder;
   }
 
+  @Nullable
+  private static FieldDescriptor toFieldDescriptor(Method method, Encodable.Field fieldAnnotation) {
+    if (method.getAnnotation(Encodable.Ignore.class) != null) {
+      return null;
+    }
+
+    String name = toGetterName(method, fieldAnnotation);
+    if (name == null) {
+      return null;
+    }
+    FieldDescriptor.Builder builder = FieldDescriptor.builder(name);
+    for (Annotation annotation : method.getAnnotations()) {
+      if (!annotation.annotationType().isAnnotationPresent(ExtraProperty.class)) {
```

Comment:
if we have Annotation Processor implemented, should we just check the annotation type is registered? in that case, we may not need `ExtraProperty`?

Reply:
we still need `ExtraProperty` for the code-gen decoders as at compile time we don't yet know what annotations will be registered at runtime.

Additionally, having this encoder implementation make assumptions about the environment it is running it would increase coupling to that environment, not something we want.

Also note that this encoder is a singleton and it caches information about types, so it can be run by differently configured encoder contexts.